fix: solve provider crash

seprintour opened the pull request:
This fix #37

pavlovcik commented on static/scripts/pay.ts:
```diff
@@ -16,10 +16,19 @@ const ErrorHandler = (error: any, extra: string | undefined = undefined) => {
+    if(error?.message?.includes("missing provider")) {
+      alert("Error: Please use a web3 enabled browser.");
+    } else {
+      alert("Error: Please connect your wallet.");
```
Please don't use alert. Instead, this should just error in the console.

seprintour replied:
> Please don't use alert. Instead, this should just error in the console.

Oh, i just thought they should see the error clearly

pavlovcik replied:
There is another active pull request which displays errors in a toaster which we should seek to integrate here once that's merged in, optimally.

pavlovcik commented on static/scripts/pay.ts:
```diff
@@ -30,11 +39,22 @@ const withdraw = async (signer: JsonRpcSigner, txData: TxType, predefined: strin
+    } else {
+      alert("Error: Please connect your wallet.");
+    }
+    return { balance: 0, allowance: 0 };
```
I see that you want to render `0` for both. Unfortunately this is made up value and is not accurate information. Instead I think it makes sense to hide the table row entirely. Or for now, it could be simpler to write "N/A" 

If types are too tedious, even a `-1` would signal an error more clearly as negative for either value is impossible.

pavlovcik commented on static/scripts/pay.ts:
```diff
@@ -73,7 +98,7 @@ export const pay = async (): Promise<void> => {
       await withdraw(signer, txData, predefined);
     } catch (error: unknown) {
-      console.error(error);
+      console.log(error);
```
It doesn't make sense to me to change this to `console.log` if you are displaying an error you should use `console.error` (assuming you want to continue function execution. Normally you should just let the error throw normally without the try/catch.

pavlovcik commented on static/scripts/pay.ts:
```diff
@@ -30,11 +39,22 @@ const withdraw = async (signer: JsonRpcSigner, txData: TxType, predefined: strin
-  const allowance = await tokenContract.allowance(txData.owner, permit2Address);
-  return { balance, allowance };
+  try {
+    const provider = new ethers.providers.Web3Provider((window as any).ethereum)
```
You should use a window object polyfill type. I thought we banned `any` in our repository!

For example

```typescript
declare global {
	interface Window {
		ethereum: MetaMaskInpageProvider;
	}
}
```

pavlovcik commented:
Would you mind waiting until #41 is merged and logging your errors to the toaster?

seprintour commented:
> Would you mind waiting until #41 is merged and logging your errors to the toaster?

Sure, a toaster would be better for displaying errors

pavlovcik commented:
Almost done, just waiting for devpanther to merge my fixes then we can merge their pull request!

https://github.com/ubiquity/pay.ubq.fi/pull/41#issuecomment-1522129862

pavlovcik commented:
It has been merged thank you for your patience @seprintour

pavlovcik commented on react-app-env.d.ts:
This seems wrong given that we are not using React.

seprintour replied:
Oh yeah, just used to naming it that way.. would change the file name to `global`

seprintour replied:
> This seems wrong given that we are not using React.

I'd revert back to the `react-app` naming, it just looks more cleaner even though it is not a react app.

its more of typescript than react

pavlovcik commented on static/scripts/pay.ts:
```diff
@@ -30,18 +39,29 @@ const withdraw = async (signer: JsonRpcSigner, txData: TxType, predefined: strin
   const trAllowance = document.querySelector(".tr-allowance") as Element;
-  trBalance.textContent = `$${ethers.utils.formatUnits(balance, 18)}`;
-  trAllowance.textContent = `$${ethers.utils.formatUnits(allowance, 18)}`;
+  trBalance.textContent = balance > 0 ? `$${ethers.utils.formatUnits(balance, 18)}` : 'N/A';
+  trAllowance.textContent = balance > 0 ? `$${ethers.utils.formatUnits(allowance, 18)}` : 'N/A';
```
```suggestion
  trBalance.textContent = balance ? `$${ethers.utils.formatUnits(balance, 18)}` : 'N/A';
  trAllowance.textContent = balance ? `$${ethers.utils.formatUnits(allowance, 18)}` : 'N/A';
```

Shouldn't that do the same thing?

seprintour replied:
Yes, it should.. Thanks

pavlovcik replied:
Now I'm also wondering how USDT and USDC render with `ethers.utils.formatUnits(allowance, 18)` but perhaps that can be a different bounty.

seprintour replied:
> trBalance.textContent = balance ? `$${ethers.utils.formatUnits(balance, 18)}` : 'N/A';

Right, USDT and USDC both have 6 decimals.. DAI has 18

seprintour replied:
> Now I'm also wondering how USDT and USDC render with `ethers.utils.formatUnits(allowance, 18)` but perhaps that can be a different bounty.

I have handled the new issue here

seprintour commented:
Fixed #49

pavlovcik commented:
> Fixed https://github.com/ubiquity/pay.ubq.fi/issues/49

You should handle this in a separate pull request. It's cleaner to associate one dedicated pull request per issue. Also our compensation system breaks down if you try and close multiple issues with a single pull request.

pavlovcik commented on static/scripts/pay.ts:
```diff
@@ -182,11 +182,14 @@ const fetchTreasury = async (): Promise<{ balance: number; allowance: number }>
-  trAllowance.textContent = balance ? `$${ethers.utils.formatUnits(allowance, 18)}` : 'N/A';
+  // get token decimals
+  const decimals = await tokenContract.decimals();
+  trBalance.textContent = balance ? `$${ethers.utils.formatUnits(balance, +decimals)}` : 'N/A';
+  trAllowance.textContent = balance ? `$${ethers.utils.formatUnits(allowance, +decimals)}` : 'N/A';
```
I'm surprised that you need to coerce `decimals` into a number. 

Shouldn't `await tokenContract.decimals();` return a number?

seprintour replied:
> I'm surprised that you need to coerce `decimals` into a number.
> 
> Shouldn't `await tokenContract.decimals();` return a number?

it is, i just thought the call might return a string

seprintour replied:
I'd branch of to a new PR, but how do i disassociate this PR from the decimal issue

pavlovcik replied:
You can try editing your original comment, and if that doesn't work, check the right side control panel on the GitHub UI
<img width="372" alt="Screenshot 2023-04-26 at 08 08 23" src="https://user-images.githubusercontent.com/4975670/234426539-4dfbf9f7-5917-4814-ad8f-8e802e6609c2.png">

seprintour replied:
> You can try editing your original comment

Done, pull request 50 should close the new issue

pavlovcik commented:
So this is all good to merge?

seprintour commented:
> So this is all good to merge?

Yup, all good

seprintour commented:
> <img alt="image" width="865" src="https://user-images.githubusercontent.com/4975670/234427933-324a9880-aaaf-4dc6-baf5-a1648317f150.png">
> 
> This still shows `loading...`

1 sec

pavlovcik commented:
I commented before testing but looks like there is still an issue.

seprintour commented:
> I commented before testing but looks like there is still an issue.
> 
> Make sure that it does not halt execution at that crash (instead you should catch it) or else the web app commit hash doesn't render (which could be very useful for debugging in the future!)

Fixed, apparently we need `balance > 0` not `balance ?` directly

seprintour commented:
Also checked if signer is valid before checking for decimals and permit

pavlovcik commented:
It's almost there but looks bad to throw the same error twice 

<img width="575" alt="image" src="https://user-images.githubusercontent.com/4975670/234429508-54c1395c-ade2-49bd-a2c8-7489cdd6f8d3.png">

seprintour commented:
> It's almost there but looks bad to throw the same error twice. Perhaps you can throw and catch at a higher level to halt execution only for the function for both of the "N/A"s rendered.
> 
> <img alt="image" width="575" src="https://user-images.githubusercontent.com/4975670/234429508-54c1395c-ade2-49bd-a2c8-7489cdd6f8d3.png">

Fixed, i moved the wallet errors to `connectWallet` only. it is called before `fetchTreasury` and should be the one to trigger such error

pavlovcik commented on static/scripts/pay.ts:
```diff
@@ -132,43 +141,52 @@ const withdraw = async (signer: JsonRpcSigner, txData: TxType, predefined: strin
-      // enable the button once on the correct network
-      enableClaimButton();
+    // watch for chain changes
+    (window as any).ethereum.on("chainChanged", async (chainId: string) => {
```
```suggestion
    window.ethereum.on("chainChanged", async (chainId: string) => {
```

pavlovcik commented:
There's some inconsistent behavior where it hangs here 

![image](https://user-images.githubusercontent.com/4975670/234486141-db3f7045-dfd7-4b5e-a1a9-eb9ece084edc.png)
